sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/orc/OrcUtils.scala

```diff
@@ -117,29 +115,16 @@ object OrcUtils extends Logging {
           }
         })
       } else {
-        if (isCaseSensitive) {
-          Some(requiredSchema.fieldNames.map { name =>
-            orcFieldNames.indexWhere(caseSensitiveResolution(_, name))
-          })
-        } else {
-          // Do case-insensitive resolution only if in case-insensitive mode
-          val caseInsensitiveOrcFieldMap =
-            orcFieldNames.zipWithIndex.groupBy(_._1.toLowerCase(Locale.ROOT))
-          Some(requiredSchema.fieldNames.map { requiredFieldName =>
-            caseInsensitiveOrcFieldMap
-              .get(requiredFieldName.toLowerCase(Locale.ROOT))
-              .map { matchedOrcFields =>
-                if (matchedOrcFields.size > 1) {
-                  // Need to fail if there is ambiguity, i.e. more than one field is matched.
-                  val matchedOrcFieldsString = matchedOrcFields.map(_._1).mkString("[", ", ", "]")
-                  throw new RuntimeException(s"""Found duplicate field(s) "$requiredFieldName": """
-                    + s"$matchedOrcFieldsString in case-insensitive mode")
-                } else {
-                  matchedOrcFields.head._2
-                }
-            }.getOrElse(-1)
-          })
+        val resolver = if (isCaseSensitive) caseSensitiveResolution else caseInsensitiveResolution
+        // Need to fail if there is ambiguity, i.e. more than one field is matched.
+        requiredSchema.fieldNames.foreach { requiredFieldName =>
+          val matchedOrcFields = orcFieldNames.filter(resolver(_, requiredFieldName))
+          if (matchedOrcFields.size > 1) {
+            throw new RuntimeException(s"""Found duplicate field(s) "$requiredFieldName": """ +
+              s"${matchedOrcFields.mkString("[", ", ", "]")} in case-insensitive mode")
```

Comment:
Actually, we do this in both mode, but `orcFieldNames` doesn't have case-sensitively duplicated names.